File: Difficulty: Medium/Frog Jump/frog-jump.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
  int helper(vector<int> &height, int n, vector<int> &dp)
{
    if (n == 0)
        return 0;
    if (dp[n] != -1)
        return dp[n];
    int left = helper(height, n - 1, dp) + abs(height[n] - height[n - 1]);
    int right = INT_MAX;
    if (n - 2 >= 0)
    {
        right = helper(height, n - 2, dp) + abs(height[n] - height[n - 2]);
    }
    return dp[n] = min(left, right);
}
int minCost(vector<int> &height)
{
    int n = height.size();
    vector<int> dp(n, -1);
    return helper(height, n - 1, dp);
}
};


//{ Driver Code Starts.

int main() {
    string str;
    getline(cin, str);
    int t = stoi(str);
    while (t--) {
        getline(cin, str);
        stringstream ss(str);
        vector<int> arr;
        int num;
        while (ss >> num) {
            arr.push_back(num);
        }
        Solution ob;
        cout << ob.minCost(arr) << endl;
        cout << "~" << endl;
    }

    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/Frog Jump/frog-jump.java
// User function Template for Java
class Solution {
    int helper(int[] height, int ind, int[] dp) {
        if (ind == 0)
            return 0;
        if (dp[ind] != -1)
            return dp[ind];
        int left = helper(height, ind - 1, dp) + Math.abs(height[ind] - height[ind - 1]);
        int right = Integer.MAX_VALUE;
        if (ind > 1)
            right = helper(height, ind - 2, dp) + Math.abs(height[ind] - height[ind - 2]);
        return dp[ind] = Math.min(left, right);

    }

    int minCost(int[] height) {
        // code here
        int n = height.length;
        int[] dp = new int[n + 1];
        Arrays.fill(dp, -1);
        return helper(height, n-1, dp);

    }
}